adr/ADR-48.md

```diff
@@ -0,0 +1,137 @@
+# JetStream Distributed Counter CRDT
+
+| Metadata | Value      |
+|----------|------------|
+| Date     | 2025-01-09 |
+| Author   | @ripienaar |
+| Status   | Approved   |
+| Tags     | jetstream  |
+
+| Revision | Date       | Author     | Info                    |
+|----------|------------|------------|-------------------------|
+| 1        | 2025-01-09 | @ripienaar | Document Initial Design |
+
+## Context and Motivation
+
+We wish to provide a distributed counter that will function in Clusters, Super Clusters, through Sources and any other way that data might reach a Stream.
+
+We will start with basic addition and subtraction primitives which will automatically behave as a CRDT and be order independent.
+
+Related:
+
+ * [Netflix Distributed Counter](https://netflixtechblog.com/netflixs-distributed-counter-abstraction-8d0c45eb66b2)
+
+## Solution Overview
+
+A Stream can opt-in to supporting Counters which will allow any subject to be a counter.
+
+Publishing a message to such a Stream will load the most recent message on the subject, increment its value and save 
+the new message with the latest value in the body. The header is preserved for downstream processing by Sources and for 
+visibility and debugging purposes.
+
+```bash
+$ nats s get COUNTER --last-for counter.hits
+Item: COUNTER#22802062 received 2025-01-09 18:05:07.93747413 +0000 UTC on Subject counter.hits
+
+Headers:
+  Nats-Incr: +2
+
+{"val":"100"}
+$ nats pub counter.hits '' -J -H "Nats-Incr:+1"
+$ nats s get COUNTER --last-for counter.hits
+Item: COUNTER#22802063 received 2025-01-09 18:06:00 +0000 UTC on Subject counter.hits
+
+Headers:
+  Nats-Incr: +1
+
+{"val":"101"}
+```
+
+## Design and Behavior
+
+The goal is to support addition and subtraction only and that a Stream that Sources 10s of other Streams all with 
+counters will effectively create a big combined counter holding totals contributed to by all sources.
+
+Handling published messages has the follow behavior and constraints:
+
+ * The header holds values like `+1`, `-1` and `-10`, in other words any valid `int64`, if the value fails to parse 
+   the message is rejected with an error. A value of `0` is valid.
+ * When publishing a message to the subject the last value is loaded, the body is parsed, incremented and written 
```

Comment:
Feel like we need to strongly define what "the last value" is, i.e. it must be from the leader only, cannot be direct-get'd from a replica.

Also do we have to think about interactions with `Nats-Expected-Last-Sequence`, `Nats-Expected-Subject-Last-Sequence` etc?

Reply:
Have marked those all as rejected and made a small clarification